enhancements/monitoring/replace-prometheus-adapter-with-metrics-server.md

```diff
@@ -0,0 +1,261 @@
+
+#### Story 3
+
+* As a cluster admin, I want a simplified monitoring architecture in OpenShift, so that I can reduce the complexity and maintenance overhead of the monitoring system.
+
+### Goals
+
+* Replace Prometheus Adapter with Metrics Server while maintaining feature parity with the existing stack 
+* Ensure no performance regression occurs when replacing Prometheus Adapter with Metrics Server
+* Improved scale up time for autoscaling
+
+### Non-Goals
+
+* Allow optional cluster monitoring stack
+* Moving `openshift-monitoring/cluster-monitoring-config` config-map to a Custom Resource
+
+## Proposal
+
+The proposal is to replace the Prometheus Adapter with the Metrics Server for monitoring resource metrics in OpenShift.
+
+### Workflow Description
+
+The workflow for using the Metrics Server in OpenShift for monitoring resource metrics would involve the following steps:
+
+1. User enables `TechPreviewNoUpgrade` featureset
+
+    ```yaml
+    apiVersion: config.openshift.io/v1
+    kind: FeatureGate
+    metadata:
+      name: cluster
+    spec:
+      featureSet: TechPreviewNoUpgrade
+    ```
+
+    With this configuration, the cluster monitoring operator should:
+    1. Deploy the `metrics-server` resources in the `openshift-monitoring` namespace
+    1. Verify that the `metrics-server` service is functional by querying the `/readyz` endpoint
+    1. Wait for `5 x metrics-server's scrape interval` to have enough data populated in `metrics-server` in-memory
+    1. Update `APIService` resource to point to `metrics-server` API instead of `prometheus-adapter` API
+    1. Delete `prometheus-adapter` resources
+
+### API Extensions
+
+No new API extensions are required. For tech preview, the only update required
+is the addition of a `MetricsServer` feature gate to the OpenShift API
+`TechPreviewNoUpgrade` feature set. See [Design Details](#design-details) for
+the specifics.
+
+### Implementation Details/Notes/Constraints
+
+#### Impact for Hosted Control Planes (HyperShift)
+
+1. The existing resource metrics API implementation, prometheus-adapter, is deployed on the guest cluster's data plane by CMO which is itself deployed by the Cluster Version operator. The same will be true for the Metrics server because it requires kubelet scraping. Consequently, we don't anticipate any effects on Hypershift as part of this enhancement.
+
+#### Configuration/Enablement
+
+Initially the Metrics Server must be explicitly enabled by the user by configuring the `TechPreviewNoUpgrade` feature set. 
+
+#### Resource Impact
+
+Once metrics-server is enabled, by default 2 replicas of `metrics-server` instances are being deployed for for multi-nodes clusters and 1 for SNO clusters in the openshift-monitoring namespace.
+
+Once metrics-server is fully functional (after 5 scrape cycles done by metrics-server) we will remove `prometheus-adapter` resources fully.
```

Comment:
Did you mean we don't need to wait 5 scrapes? I see `initialDelaySeconds` is added to readiness probe which waits till 2 scrapes are done

Reply:
I am not entirely sure how you were planning to implement it, but just waiting for metrics-server to report ready on its `readyz` endpoint should be good enough since we already have internal logic there to make sure that it is ready to serve metrics.

Reply:
I think that I already suggested it previously so :+1: for me. It should be enough for CMO to wait for the metrics-server deployment to be ready I believe.